// @saberhq/token-utils/dist/esm/tokenOwner.js
import { __awaiter } from "tslib";
import { ASSOCIATED_TOKEN_PROGRAM_ID, getATAAddress, SPLToken, TOKEN_PROGRAM_ID, } from ".";
/**
 * Wrapper around a token account owner to create token instructions.
 */
export class TokenOwner {
    constructor(owner) {
        this.owner = owner;
    }
    /**
     * Gets the user's ATA.
     * @param mint
     * @returns
     */
    getATA(mint) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield getATAAddress({ mint, owner: this.owner });
        });
    }
    /**
     * Transfers tokens to a token account.
     * @param amount Amount of tokens to transfer.
     * @param to Token account to transfer to.
     * @returns The transaction instruction.
     */
    transfer(amount, to) {
        return __awaiter(this, void 0, void 0, function* () {
            return SPLToken.createTransferInstruction(TOKEN_PROGRAM_ID, yield this.getATA(amount.token.mintAccount), to, this.owner, [], amount.toU64());
        });
    }
    /**
     * Transfers tokens to a token account, checked..
     * @param amount Amount of tokens to transfer.
     * @param to Token account to transfer to.
     * @returns The transaction instruction.
     */
    transferChecked(amount, to) {
        return __awaiter(this, void 0, void 0, function* () {
            return SPLToken.createTransferCheckedInstruction(TOKEN_PROGRAM_ID, yield this.getATA(amount.token.mintAccount), amount.token.mintAccount, to, this.owner, [], amount.toU64(), amount.token.decimals);
        });
    }
    /**
     * Mints tokens to a token account.
     * @param amount Amount of tokens to transfer.
     * @param to Token account to transfer to.
     * @returns The transaction instruction.
     */
    mintTo(amount, to) {
        return SPLToken.createMintToInstruction(TOKEN_PROGRAM_ID, amount.token.mintAccount, to, this.owner, [], amount.toU64());
    }
    /**
     * Creates an associated token account instruction.
     * @param mint Mint of the ATA.
     * @param payer Payer to create the ATA. Defaults to the owner.
     * @returns The transaction instruction.
     */
    createATA(mint, payer = this.owner) {
        return __awaiter(this, void 0, void 0, function* () {
            return SPLToken.createAssociatedTokenAccountInstruction(ASSOCIATED_TOKEN_PROGRAM_ID, TOKEN_PROGRAM_ID, mint, yield this.getATA(mint), this.owner, payer);
        });
    }
}
